Blog post about the implementation of job queue with Redis (#1540)
quarkusio/quarkusio.github.io: _posts/2022-09-15-redis-job-queue.adoc

image::/assets/images/posts/redis-job-queue/pattern.png[align="center"]

There are plenty of variants of that pattern, so let's focus on the following application.
We have an application managing heroes and villains.
The application offers the possibility to simulate a fight between a random hero and a random villain.
The fight simulation is delegated to _fight simulators_, applications dedicated to that task.

image::/assets/images/posts/redis-job-queue/application.png[align="center"]

In this context, the main application submits the fight request to the job queue.
Then, the fight simulators poll the submitted fight request and execute them.

The fight outcomes are communicated using another Redis feature: pub/sub communication.
The simulators send the outcome to a _channel_ consumed by the application.
The application then broadcasts these outcomes to a web page.

This post only discusses the interaction with Redis.
The rest of the application is straightforward and just uses RESTEasy Reactive and Hibernate ORM with Panache.
You can find the full code of the application on https://github.com/cescoffier/quarkus-redis-job-queue-demo.

== Submitting jobs

> The first task is to modelize the job queue.

Reviewer: ```suggestion
The first task is to model the job queue.
```

We are using a https://redis.io/docs/data-types/#lists[Redis list] to store the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightRequest.java[FightRequest].

[source,java]
----
package me.escoffier.quarkus.redis.fight;

public record FightRequest(String id, Hero hero, Villain villain) {

}
----

Redis lists distinguish the left side of the list from the right side of the list.

> This distinction allows implementing a FIFO queue where we write on the left side and consume from the right side.

Reviewer: Maybe instead of `left` and `right` we should use `beginning` and `end`? I find the former rather confusing.

_EDIT:_ I guess you are using these terms because that is what Redis uses - in that case, I would say you made the correct decision

Author: Yes, Redis uses left and right when dealing with redis list.
`BRPOP` stands for _blocking right pop_ .... We can also debate about the blocking aspect ;-)

To manipulate a Redis list, we need the _group of commands_ associated with this data structure.
In the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/supes-application/src/main/java/me/escoffier/quarkus/redis/supes/SupesService.java[SupesService class], we inject the `RedisDataSource` and retrieve the group of commands:

[source,java]
----
public SupesService(RedisDataSource dataSource, ...) {
    commands = dataSource.list(FightRequest.class);
  // ...
}
----

Let's now look at the `submitAFight` method:

[source,java]
----
public FightRequest submitAFight() {
    var hero = Hero.getRandomHero();
    var villain = Villain.getRandomVillain();
    var id = UUID.randomUUID().toString();
    var request = new FightRequest(id, hero, villain);
    commands.lpush("fight-requests", request);
    return request;
}
----

> The `submitaFight` method retrieves the random fighters, computes an id,  builds the `FightRequest` instance, and executes the `LPUSH` command.

Reviewer: ```suggestion
The `submitAFight` method retrieves the random fighters, computes an id,  builds the `FightRequest` instance, and executes the `LPUSH` command.
```

The `LPUSH`  command writes the given item to the left side of the list stored at the given key (`fight-requests`).

== Receiving the job requests

Let's now look at the other side: the fight simulators.

> The simulators poll the `FightRequests` from the Redis list modelizing our job queue and simulate the fight.

Reviewer: ```suggestion
The simulators poll the `FightRequests` from the Redis list that models our job queue and simulate the fight.
```

The simulator is implemented in https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightSimulator.java[`me.escoffier.quarkus.redis.fight.FightSimulator`].
The constructor receives a configured name (to distinguish multiple simulators) and the Redis data source.
It creates the objects to emit the Redis commands to read from a Redis list:

[source,java]
----

> public FightSimulator(@ConfigProperty(name = "simulator-name") String name, Logger logger, RedisDataSource ds) {

Reviewer: Logger in the constructor? 
It's valid, but it's not really something use often :)

Author: Initially, I was logging in to the constructor, which requires constructor injection, because the field was not yet injected.

    this.name = name;
    this.logger = logger;
    this.queue = ds.list(FightRequest.class);
    // ...
}
----

The simulator polls the fight requests and for each of them simulate the fight.
The implementation is an _infinite_ loop (it only stops when the application is shut down).
In each iteration, it reads the pending `FightRequest` from the right side of the queue with the `BRPOP` command.
If there is no pending request, it restarts from the beginning of the loop.
If it has a request, it simulates the fight:

[source,java]
----
@Override
public void run() {
    logger.infof("Simulator %s starting", name);
    while ((!stopped)) {
        KeyValue<String, FightRequest> item =
            queue.brpop(Duration.ofSeconds(1), "fight-requests");
        if (item != null) {
            var request = item.value();
            var result = simulate(request);
            //...
        }

>     }

Reviewer: Not related to this blog post, but maybe we should have some kind of super easy to do this, like an annotation on a method that will be called by Quarkus - something like what Reactive Messaging does.

Reviewer: 👍🏼

        if (item != null) {
            var request = item.value();
            var result = simulate(request);
            publisher.publish("fight-results", result);  // Send the outcome
           }
    }
}
----

== Receiving the fight outcome

At that point:

* we submit the fight request into the job queue,
* we consume that queue and simulate the fight,
* we send the outcome to the `fight-results` channel.

So, the only missing piece is the consumption of that channel.
Let's return to the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/supes-application/src/main/java/me/escoffier/quarkus/redis/supes/SupesService.java[`me.escoffier.quarkus.redis.supes.SupesService`] class.
In the constructor, we also inject the `ReactiveRedisDataSource`, the reactive variant of the Redis data source.
Then, in the constructor code, we subscribe to the `fight-results`.

[source,java]
----
public SupesService(RedisDataSource dataSource, ReactiveRedisDataSource reactiveRedisDataSource) {
    commands = dataSource.list(FightRequest.class);
    stream = reactiveRedisDataSource.pubsub(FightResult.class).subscribe("fight-results")

>             .broadcast().toAllSubscribers();

Reviewer: I think this needs more explaining

The fight won't happen as we have no simulator.
However, the fight requests have been submitted and stored in the list.
So they are not lost.

Now, in the second terminal, navigate to the `fight-simulator` directory, and run:

[source,shell]
----
mvn package
java -Dsimulator-name=A -Dquarkus.redis.hosts=redis://localhost:53853 -jar target/quarkus-app/quarkus-run.jar
----

*IMPORTANT*: update the `quarkus.redis-hosts` with the one copied above.

As soon as you start it, it processes the pending fight requests:

[source,text]
----
2022-09-11 15:31:58,914 INFO  [me.esc.qua.red.fig.FightSimulator] (Thread-3) Simulator A is going to simulate a fight between Pakku and Tulon Voidgazer
2022-09-11 15:31:59,786 INFO  [me.esc.qua.red.fig.FightSimulator] (Thread-3) Simulator A is going to simulate a fight between Comet Zuko and Arishem The Judge (Knullified)
2022-09-11 15:32:01,809 INFO  [me.esc.qua.red.fig.FightSimulator] (Thread-3) Simulator A is going to simulate a fight between Ms. America and Kazumi (Devil Form)
----

If you go back to the web page, the winners get a _halo_:

image::/assets/images/posts/redis-job-queue/screenshot-winner.png[align="center"]

> Now, in the thrid terminal, navigate to the `fight-simulator` directory, and run:

Reviewer: ```suggestion
Now, in the third terminal, navigate to the `fight-simulator` directory, and run:
```

[source,shell]
----
java -Dsimulator-name=B -Dquarkus.redis.hosts=redis://localhost:53853 -jar target/quarkus-app/quarkus-run.jar
----

*IMPORTANT*: as in the previous command, update the `quarkus.redis-hosts` with the one copied above.

Go back to the web page and click on the `fight!` button a few times.
Check the logs of both simulators to see that the fight requests are now dispatched beween the two simulators.

== Summary

This posts explains how you can implement a job queue with Redis and the Quarkus Redis datasource API.

> It also presents how to use Redis pub/sub.

Reviewer: I don't think this sentence adds much value.
